nixos/modules/virtualisation/cri-o.nix

```diff
@@ -80,60 +85,65 @@ in
-    environment.etc."cni/net.d/99-loopback.conf".source = copyFile "${pkgs.cri-o-unwrapped.src}/contrib/cni/99-loopback.conf";
-
-    # Enable common /etc/containers configuration
-    virtualisation.containers.enable = true;
-
-    systemd.services.crio = {
-      description = "Container Runtime Interface for OCI (CRI-O)";
-      documentation = [ "https://github.com/cri-o/cri-o" ];
-      wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      path = [ cfg.package ];
-      serviceConfig = {
-        Type = "notify";
-        ExecStart = "${cfg.package}/bin/crio";
-        ExecReload = "/bin/kill -s HUP $MAINPID";
-        TasksMax = "infinity";
-        LimitNOFILE = "1048576";
-        LimitNPROC = "1048576";
-        LimitCORE = "infinity";
-        OOMScoreAdjust = "-999";
-        TimeoutStartSec = "0";
-        Restart = "on-abnormal";
+  config = mkMerge [
+    (mkIf cfg.enable {
+      environment.systemPackages = [ cfg.package pkgs.cri-tools ];
+
+      environment.etc."crictl.yaml".source = copyFile "${pkgs.cri-o-unwrapped.src}/crictl.yaml";
+
+      environment.etc."crio/crio.conf.d/00-default.conf".text = ''
+        [crio]
+        storage_driver = "${cfg.storageDriver}"
+
+        [crio.image]
+        ${optionalString (cfg.pauseImage != null) ''pause_image = "${cfg.pauseImage}"''}
+        ${optionalString (cfg.pauseCommand != null) ''pause_command = "${cfg.pauseCommand}"''}
+
+        [crio.network]
+        plugin_dirs = ["${pkgs.cni-plugins}/bin/"]
+        ${optionalString (cfg.networkDir != null) ''network_dir = "${cfg.networkDir}"''}
+
+        [crio.runtime]
+        cgroup_manager = "systemd"
+        log_level = "${cfg.logLevel}"
+        manage_ns_lifecycle = true
+        pinns_path = "${cfg.package}/bin/pinns"
+        hooks_dir = []
+
+        ${optionalString (cfg.runtime != null) ''
+        default_runtime = "${cfg.runtime}"
+        [crio.runtime.runtimes]
+        [crio.runtime.runtimes.${cfg.runtime}]
+        ''}
+      '';
+
+      # Enable common /etc/containers configuration
+      virtualisation.containers.enable = true;
+
+      systemd.services.crio = {
+        description = "Container Runtime Interface for OCI (CRI-O)";
+        documentation = [ "https://github.com/cri-o/cri-o" ];
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        path = [ cfg.package ];
+        serviceConfig = {
+          Type = "notify";
+          ExecStart = "${cfg.package}/bin/crio";
+          ExecReload = "/bin/kill -s HUP $MAINPID";
+          TasksMax = "infinity";
+          LimitNOFILE = "1048576";
+          LimitNPROC = "1048576";
+          LimitCORE = "infinity";
+          OOMScoreAdjust = "-999";
+          TimeoutStartSec = "0";
+          Restart = "on-abnormal";
+        };
       };
-    };
-  };
+    })
+
+    (mkIf (cfg.enable && cfg.networkDir == null) {
```

Comment:
Does installing these need to be conditional so they won't be picked up or is setting `network_dir` sufficient?

Reply:
My idea was to install them only if the network dir is default. Usually if users want to specify their own network dir then this will happen on purpose that they use a custom setup method like:
1. Start CRI-O with a custom network dir which does not contain any network
2. Use a CNI bootstrap container like flannel which creates the config
3. CRI-O picks-up the config from the network dir and uses it to create sandboxes from it.

We probably want to add the 99-loopback.conf in any case… :thinking: I'll change that.